File: app/src/main/java/com/solt9029/editmasterandroid/viewmodel/ScoreListViewModel.java
package com.solt9029.editmasterandroid.viewmodel;

import com.solt9029.editmasterandroid.model.Score;
import com.solt9029.editmasterandroid.repository.ScoreRepository;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import androidx.annotation.Nullable;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class ScoreListViewModel extends ViewModel {
    public MutableLiveData<Resource<List<Score>>> resource = new MutableLiveData<>(new Resource<>());
    public MutableLiveData<Boolean> isRefreshing = new MutableLiveData<>(false);
    public LiveEvent<Integer> navigateToScoreActivity = new LiveEvent<>();
    public MutableLiveData<String> keyword = new MutableLiveData<>();
    private CompositeDisposable compositeDisposable = new CompositeDisposable();
    private ScoreRepository repository;

    @Inject
    ScoreListViewModel(ScoreRepository repository) {
        this.repository = repository;
        onLoad();
    }

    @Override
    public void onCleared() {
        compositeDisposable.clear();
    }

    public void onRefresh() {
        Timber.d("onRefresh");
        compositeDisposable.clear();

        isRefreshing.setValue(true);
        resource.setValue(Resource.startLoading(getData()));
        Disposable disposable = fetchScoreTimeline().subscribe(
                result -> {
                    isRefreshing.setValue(false);
                    resource.setValue(Resource.finishLoadingSuccess(result));
                    Timber.d("refresh success");
                },
                error -> {
                    resource.setValue(Resource.finishLoadingFailure(error));
                    Timber.d("refresh failure");
                }
        );
        compositeDisposable.add(disposable);
    }

    public void onLoad() {
        compositeDisposable.clear();

        resource.setValue(Resource.startLoading(null));
        Disposable disposable = fetchScoreTimeline().subscribe(
                result -> resource.setValue(Resource.finishLoadingSuccess(result)),
                error -> resource.setValue(Resource.finishLoadingFailure(error))
        );
        compositeDisposable.add(disposable);
    }

    public void onLoadMore() {
        if (getIsLoading()) {
            return;
        }

        Integer maxId = getMaxId();
        if (maxId != null && maxId < 1) {
            return;
        }

        resource.setValue(Resource.startLoading(getData()));
        Disposable disposable = fetchScoreTimeline(maxId).subscribe(
                result -> resource.setValue(Resource.finishLoadingSuccess(addData(result))),
                error -> resource.setValue(Resource.finishLoadingFailure(error))
        );
        compositeDisposable.add(disposable);
    }

    private Integer getMaxId() {
        Integer maxId = null;
        List<Score> data = getData();
        if (data != null) {
            maxId = data.get(data.size() - 1).getId() - 1;
        }
        return maxId;
    }

    @Nullable
    public List<Score> getData() {
        List<Score> data = null;
        if (resource.getValue() != null) {
            data = resource.getValue().data;
        }
        return data;
    }

    private boolean getIsLoading() {
        boolean isLoading = true;
        if (resource.getValue() != null) {
            isLoading = resource.getValue().isLoading;
        }
        return isLoading;
    }

    private Single<List<Score>> fetchScoreTimeline() {
        return fetchScoreTimeline(null);
    }

    private Single<List<Score>> fetchScoreTimeline(Integer maxId) {
        return repository.getScoreTimeline(null, keyword.getValue(), maxId, null)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    private List<Score> addData(List<Score> result) {
        List<Score> newData = new ArrayList<>();
        if (getData() != null) {
            newData.addAll(getData());
        }
        newData.addAll(result);
        return newData;
    }

    public void navigateToScoreActivity(int id) {
        navigateToScoreActivity.call(id);
    }
}

File: app/src/main/java/com/solt9029/editmasterandroid/viewmodel/ScoreListViewModel.kt
package com.solt9029.editmasterandroid.viewmodel

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.solt9029.editmasterandroid.entity.Score
import com.solt9029.editmasterandroid.repository.ScoreRepository
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import java.util.*
import javax.inject.Inject

class ScoreListViewModel @Inject constructor(
        private val repository: ScoreRepository
) : ViewModel() {

    var resource = MutableLiveData(Resource<List<Score>>())
    var isRefreshing = MutableLiveData(false)
    var navigateToScoreActivity = LiveEvent<Int>()
    var keyword = MutableLiveData<String?>()
    private val compositeDisposable = CompositeDisposable()

    init {
        onLoad()
    }

    public override fun onCleared() {
        compositeDisposable.clear()
    }

    fun onRefresh() {
        Timber.d("onRefresh")
        compositeDisposable.clear()

        isRefreshing.value = true
        resource.value = Resource.startLoading(resource.value?.data)
        val disposable = fetchScoreTimeline().subscribe(
                { result ->
                    isRefreshing.value = false
                    resource.value = Resource.finishLoadingSuccess(result)
                    Timber.d("refresh success")
                },
                { error ->
                    isRefreshing.value = false
                    resource.value = Resource.finishLoadingFailure(error)
                    Timber.e("refresh failure")
                    Timber.e(error.message)
                }
        )
        compositeDisposable.add(disposable)
    }

    fun onLoad() {
        Timber.d("onLoad")
        compositeDisposable.clear()

        resource.value = Resource.startLoading(null)
        val disposable = fetchScoreTimeline().subscribe(
                { result ->
                    resource.value = Resource.finishLoadingSuccess(result)
                    Timber.d("load success")
                },
                { error ->
                    resource.value = Resource.finishLoadingFailure(error)
                    Timber.e("load failure")
                    Timber.e(error.message)
                }
        )
        compositeDisposable.add(disposable)
    }

    fun onLoadMore() {
        if (resource.value?.isLoading == true) {
            return
        }

        if (maxId != null && maxId!! < 1) {
            return
        }

        Timber.d("onLoadMore")

        resource.value = Resource.startLoading(resource.value?.data)
        val disposable = fetchScoreTimeline(maxId).subscribe(
                { result ->
                    val newData = ArrayList<Score>()
                    resource.value?.data?.let { data ->
                        newData.addAll(data)
                    }
                    newData.addAll(result)
                    resource.setValue(Resource.finishLoadingSuccess(newData))
                    Timber.d("load more success")
                },
                { error ->
                    resource.value = Resource.finishLoadingFailure(error)
                    Timber.e("load more failure")
                    Timber.e(error.message)
                }
        )
        compositeDisposable.add(disposable)
    }

    private fun fetchScoreTimeline(maxId: Int? = null): Single<List<Score>> {
        return repository.getScoreTimeline(null, keyword.value, maxId, null)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
    }

    private val maxId: Int?
        get() {
            resource.value?.data?.let { data ->
                data[data.size - 1].id?.let { id ->
                    return id - 1
                }
            }
            return null
        }

    fun navigateToScoreActivity(id: Int) {
        navigateToScoreActivity.call(id)
    }
}